File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    int carry = 0;
    ListNode p1 = l1;
    ListNode p2 = l2;
    ListNode result = new ListNode();
    ListNode p = result;
    while (carry > 0 || p1 != null || p2 != null) {
      if (p1 != null && p2 != null) {
        int sum = carry + p1.val + p2.val;
        p.next = new ListNode(sum%10);
        carry = sum / 10;
        p1 = p1.next;
        p2 = p2.next;
        p = p.next;
      } else if (p1 == null && p2 == null) {
        p.next = new ListNode(carry);
        break;
      } else if (carry == 0) {
        p.next = (p1 == null) ? p2 : p1;
        break;
      } else {
        if (p1 == null) {
          p1 = p2;
          p2 = null;
        } else {
          int sum = carry + p1.val;
          p.next = new ListNode(sum%10);
          carry = sum / 10;
          p1 = p1.next;
          p = p.next;
        }
      }
    }
    return result.next;
  }
}

File: 2-add-two-numbers/2-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode* p1 = l1;
    ListNode* p2 = l2;
    ListNode result;
    ListNode* p = &result;
    int carry = 0;
    while (carry || p1 || p2) {
      if (p1 && p2) {
        int sum = p1->val + p2->val + carry;
        p->next = new ListNode(sum%10);
        p1 = p1->next;
        p2 = p2->next;
        p = p->next;
        carry = sum / 10;
      } else if (carry) {
        int sum = carry + (p1 ? p1->val : 0) + (p2 ? p2->val : 0);
        p->next = new ListNode(sum%10);
        if (p1) p1 = p1->next;
        if (p2) p2 = p2->next;
        p = p->next;
        carry = sum / 10;
      } else if (p1) {
        p->next = p1;
        p1 = nullptr;
      } else {
        p->next = p2;
        p2 = nullptr;
      }
    }
    return result.next;    
  }
};